cogs5e/models/automation/effects/damage.py

```diff
@@ -91,9 +96,20 @@ def run(self, autoctx):
         # Disable critical damage in saves (#1556)
         in_crit = (autoctx.in_crit or crit_arg) and not (nocrit or autoctx.in_save)
         if in_crit:
-            dice_ast = d20.utils.tree_map(utils.crit_mapper, dice_ast)
+            if crit_damage_type == CritDamageType.MAX_ADD:
+                dice_ast = d20.utils.tree_map(utils.max_add_crit_mapper, dice_ast)
+            elif crit_damage_type == CritDamageType.DOUBLE_ALL:
+                dice_ast.roll = d20.ast.BinOp(d20.ast.Parenthetical(dice_ast.roll), "*", d20.ast.Literal(2))
+            elif crit_damage_type == CritDamageType.DOUBLE_DICE:
+                dice_ast = d20.utils.tree_map(utils.double_dice_crit_mapper, dice_ast)
+            else:
+                dice_ast = d20.utils.tree_map(utils.crit_mapper, dice_ast)
             if critdice and not autoctx.is_spell:
-                utils.critdice_tree_update(dice_ast, int(critdice))
+                if crit_damage_type in (CritDamageType.DOUBLE_ALL, CritDamageType.DOUBLE_DICE):
+                    crit_ast = utils.crit_dice_gen(dice_ast, critdice)
+                    dice_ast.roll = d20.ast.BinOp(dice_ast.roll, "+", crit_ast)
```

Comment:
Currently the return type of `utils.crit_dice_gen` is `Optional[d20.ast.Dice]`(because of utils+L95); this will cause errors if it returns None